File: rust/src/s0128.rs
pub struct Solution {}

impl Solution {
    pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
        use std::collections::HashSet;
        // let set: HashSet<i32> = HashSet::from_iter(nums.into_iter());

        let set: HashSet<i32> =
            nums.into_iter().fold(HashSet::new(), |mut set, num| {
                set.insert(num);
                set
            });

        let mut result = 0;
        for &v in set.iter() {
            if !set.contains(&(v - 1)) {
                let mut length = 1;
                while set.contains(&(v + length)) {
                    length += 1;
                }
                result = std::cmp::max(result, length)
            }
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        println!(
            "{:?}",
            Solution::longest_consecutive(vec![0, 3, 7, 2, 5, 8, 4, 6, 0, 1])
        );
    }
}

File: cpp/s0128.cc
class Solution {
public:
  int longestConsecutive(vector<int>& nums) {
    unordered_set<int> set(nums.begin(), nums.end());
    int res = 0;
    for(auto& n : set) {
      if(!set.count(n - 1)) {
        // start count
        int len = 1;
        int curr = n + 1;
        while(set.count(curr)) {
          len += 1;
          curr += 1;
        }
        res = max(res, len);
      }
    }

    return res;
  }
};
